odoo/sql_db.py

```diff
@@ -100,6 +101,64 @@ def check(f, self, *args, **kwargs):
         raise psycopg2.OperationalError('Unable to use a closed cursor.')
     return f(self, *args, **kwargs)
 
+class Savepoint:
+    """ Reifies an active breakpoint, allows :meth:`BaseModel.savepoint` users
+    to internally rollback the savepoint (as many times as they want) without
+    having to implement their own savepointing, or triggering exceptions.
+
+    While the savepoint name is created during (and accessible after)
+    initialisation, the actual savepoint is only defined in ``__enter__``, and
+    only valid for the context-managed span. The savepoint is automatically
+    released ("committed") on context exit, unless an exception was raised in
+    which case it's rolled back first.
+
+    It can also be covered by `closing`, with similar semantics to
+    :class:`Cursor`: closing a savepoint will unconditionally roll it back
+    before releasing it.
+
+    The savepoint can safely be closed during the activation span.
+    """
+    def __init__(self, cr):
+        self.name = uuid.uuid1().hex
+        self._name = Identifier(self.name)
+        self._cr = cr
+        self._closed = False
+
+    def __enter__(self):
+        self._cr.execute(SQL('SAVEPOINT {}').format(self._name))
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close(rollback=exc_type is not None)
+
+    def close(self, *, rollback=True):
+        if not self._closed:
+            if rollback:
+                self.rollback()
+            self._cr.execute(SQL('RELEASE SAVEPOINT {}').format(self._name))
+            self._closed = True
+
+    def rollback(self):
+        self._cr.execute(SQL('ROLLBACK TO SAVEPOINT {}').format(self._name))
+
+@contextmanager
+def flushing(cr, cm):
```

Comment:
Unless you see opportunities to reuse this in more than one place, I don't see the point of factoring this out.  To me it looks like you move the logic outside of the class where it makes sense.

Reply:
Yeah it may have made more sense in the original (incorrect) version where the flushing was nested inside. When that was fixed I probably kept updating the CM instead of just dropping it.